### DISCUSSION TITLE
Is it possible to upload a file with Chrome on Android device ?

### DISCUSSION THREAD

**Discussion by Nosetiduno** [webdriverio/webdriverio#10518] [Q&A]
Hi everyone,

 I'm doing a POC with WebdriverIO and I want to check that I can upload a file on web app using Chrome browser on an android device.
For the moment, I ve managed to indentify the input, display it and set its value with the path of the file to upload :

browser.execute(function(){
        document.getElementById("inputElem").style.display = 'block';
        });

    const inputDiv = $('#inputElem');
    await inputDiv.waitForDisplayed();

let pathImage = "/storage/emulated/0/Pictures/IMG_20230606_085741.jpg";
const filePath = path.join(cwd(), pathImage);
await locatorInputUpload.setValue(filePath);

When I use this method, I use the path of the file I want to upload which is in my WebdriverIO project and I get an error page : "Something whent wrong while displaying this webpage"

 If I put the path of an image on the android device, Ive got an error saying that I have to use absolute path.

Is it possible to do this upload ? and if it is possible, the file to upload must be on the device or can be in the webdriverIO project ?

**Comment by portyu9** [comment]
Hi @Nosetiduno 

Yes, but the important distinction is **runner filesystem vs Android-device filesystem**.

This:

```js
const filePath = path.join(cwd(), '/storage/emulated/0/Pictures/...')
```

does not turn an Android path into a device path. `cwd()` refers to the machine running WebdriverIO.

If the source file is in your WDIO project, first copy it to the Android device with Appium:

```js
import fs from 'node:fs'

const localPath = './test/files/image.jpg'
const devicePath = '/sdcard/Download/image.jpg'

await browser.pushFile(
    devicePath,
    fs.readFileSync(localPath).toString('base64')
)
```

You can then try supplying the **device path** to the real `<input type="file">`:

```js
const input = await $('input[type="file"]')
await input.setValue(devicePath)
```

Be aware that mobile Chrome/Chromedriver has had filesystem-access restrictions around HTML file inputs. If a valid device path still produces `ERR_ACCESS_DENIED` or the Android file picker opens instead, that's a Chrome/Android behavior rather than something `path.join()` can fix. In that case, switch to native context and automate Android's file picker.

So a project fixture can be used, but it first needs to become accessible on the device if you're following the device-path approach.

I reviewed the complete code/error in the discussion; there were no screenshots, links, or replies.

Please mark this answer as accepted if it helped, thank you.
